src/PAModel/Serializers/SourceSerializer.cs

```diff
@@ -795,5 +798,62 @@ private static BuildVerJson GetBuildDetails()
+        {
+            foreach (var child in node.Children)
+            {
+                // group, grouContainer, gallery etc. have nested controls so run the validation for all the children.
+                if (child.Children?.Count > 0)
+                {
+                    foreach (var child1 in child.Children)
+                    {
+                        ValidateIfTemplateExists(app, child1, root, errors);
+                    }
+                }
+
+                CombinedTemplateState templateState;
+                app._templateStore.TryGetTemplate(child.Name.Kind.TypeName, out templateState);
+
+                // Some of the child components don't have a template eg. TestStep, so we can safely continue if we can't find an entry in the templateStore.
+                if (templateState == null)
+                {
+                    continue;
+                }
+
+                // If its a widget template then there must be a xml file in the pkgs directory.
+                if (templateState.IsWidgetTemplate)
+                {
+                    if (!app._templates.UsedTemplates.Any(x => x.Name == child.Name.Kind.TypeName))
+                    {
+                        errors.ValidationError(root.SourceSpan.GetValueOrDefault(), $"Widget control template: {templateState.Name}, version {templateState.Version} was not found in the pkgs directory and is referred in {root.Name.Identifier}. " +
+                            $"If the template was deleted intentionally please make sure to update the source files to remove the references to this template.");
+                    }
+                    continue;
+                }
+                // if its a component template then check if the template exists in the Src/Components directory
+                else if (templateState.IsComponentTemplate == true)
+                {
+                    if (!app._components.Keys.Any(x => x == child.Name.Kind.TypeName))
+                    {
+                        errors.ValidationError(root.SourceSpan.GetValueOrDefault(), $"Component template: {templateState.Name} was not found in Src/Components directory and is referred in {root.Name.Identifier}. " +
+                            $"If the template was deleted intentionally please make sure to update the source files to remove the references to this template.");
+                    }
+                    continue;
+                }
+                // PCF are dynamically imported controls and their template definition is stored in the DynamicControlDefinitionJson property, check if that exists.
+                else if (templateState.Id.StartsWith("http://microsoft.com/appmagic/powercontrol"))
+                {
+                    if (!templateState.ExtensionData.ContainsKey("DynamicControlDefinitionJson") || templateState.ExtensionData["DynamicControlDefinitionJson"] == null)
```

Comment:
>DynamicControlDefinitionJson [](http://example.com/codeflow?start=66&length=28)

maintenance nit: can we use a nameof() expression here
(in general, we should avoid string constants like this)